src/circuit/boolean.rs

```diff
@@ -375,35 +375,44 @@ impl Boolean {
         where E: Engine,
               CS: ConstraintSystem<E>
     {
-        let c = Self::xor(&mut cs, a, b)?;
-
-        match c {
-            Boolean::Constant(false) => {
-                Ok(())
-            },
-            Boolean::Constant(true) => {
-                Err(SynthesisError::Unsatisfiable)
+        match (a, b) {
+            (&Boolean::Constant(a), &Boolean::Constant(b)) => {
+                if a == b {
+                    Ok(())
+                } else {
+                    Err(SynthesisError::Unsatisfiable)
+                }
             },
-            Boolean::Is(ref res) => {
+            (&Boolean::Constant(true), a) | (a, &Boolean::Constant(true)) => {
                 cs.enforce(
-                    || "enforce equals zero",
+                    || "enforce equal to one",
                     |lc| lc,
                     |lc| lc,
-                    |lc| lc + res.get_variable()
+                    |lc| lc + CS::one() - &a.lc(CS::one(), E::Fr::one())
                 );
 
                 Ok(())
             },
-            Boolean::Not(ref res) => {
+            (&Boolean::Constant(false), a) | (a, &Boolean::Constant(false)) => {
                 cs.enforce(
-                    || "enforce equals one",
+                    || "enforce equal to zero",
                     |lc| lc,
                     |lc| lc,
-                    |lc| lc + CS::one() - res.get_variable()
+                    |_| a.lc(CS::one(), E::Fr::one())
                 );
 
                 Ok(())
             },
+            (a, b) => {
+                cs.enforce(
+                    || "enforce equal",
+                    |_| a.lc(CS::one(), E::Fr::one()),
+                    |lc| lc + CS::one(),
+                    |_| b.lc(CS::one(), E::Fr::one())
```

Comment:
Sorry! Another one of those "strictly better" nits. Equality of `a` and `b` is always better as `0 * 0 = a - b`.

By the way, I'm sure you could find examples of me not following my own advice elsewhere in the circuit code. :P